[doc] Add policy for merging and promotion

gavindidrichsen opened the pull request:
Graham Wilson identified a current problem with core-plan reviews [on slack](https://habitat-sh.slack.com/archives/CGS0MA3HV/p1579652201003400): it often takes a long time to get a PR reviewed, merged, and promoted.  His proposal is in effect that for certain 'veteran' core-plan contributors and for a subset of plans, the veteran should be able to merge and promote his plan if after some time (7+ days) there has been no feedback or movement.

This is a basic policy draft document.  Please add any more detail or questions to additionally answer.

stevendanna commented on docs/dev/policy_documents/merging-and-promoting-conventions.md:
```diff
@@ -0,0 +1,40 @@
+## Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+What plan is eligable for the fast-track criteria:
```
```suggestion
What pull requests are eligible for the fast-track criteria:
```

gavindidrichsen replied:
Thanks

stevendanna commented on docs/dev/policy_documents/merging-and-promoting-conventions.md:
```diff
@@ -0,0 +1,40 @@
+## Motivation
+[motivation]: #motivation
+
+If someone who contributes heavily to core plans, has a pull request for a minor change open for 7days +, with all tests passing, it would be good if the same person can also merge and promote if there is a lack of peer reviews. One contributor (Graham Wilson, predominant) has hundreds of updates in his backlog, but has been holding back because of delays in the review process.
```
```suggestion
Our code review process is currently slow, frustrating core-plan maintainers who are donating their personal time or their employer's time to the project. One contributor (Graham Wilson, predominant) has hundreds of updates in his backlog, but has been holding back because of delays in the review process.

If someone who contributes heavily to core plans has a pull request (PR) for a minor change open for 7+ days, the automated testing already in place and the small scope of the change makes the limited value of additional code-review small compared to the cost it is imposing on some maintainers.  In this situation, allowing veteran maintainers to use their own judgement and immediately merge the PR will help ensure core-plan maintainers can make the most of the time they spend on this project.
```

gavindidrichsen replied:
Perfect!

stevendanna commented on docs/dev/policy_documents/merging-and-promoting-conventions.md:
```diff
@@ -0,0 +1,40 @@
+
+* Only identified 'veteran' core-plan contributor's have permission to use this fast-track policy.  This list will be agreed and specified in the MAINTAINERS.md
+* This fasttrack will only apply to non-baseplan PRs and PRs that do not have a label 'DO-NOT-MERGE' or are otherwise marked for caution.
+* If no feedback has been added to the PR after 7+ days, then the veteran contributor can merge and promote the PR.
```
Suggested Addition:

* The PR consists of minor changes. For example, version bumps, style improvements, bug fixes that don't change the scope of the plan.

Minor style suggestion: 

I think it might be better to rephrase these eligibility requirements in the form of positive assertions that can be checked:

- PR was opened by 'veteran' core-plan contributor. Veteran core-plan contributors are specified in MAINTAINERS.md;
- PR is for a non-base plan and does not have the 'DO-NOT-MERGE' label;
- All automated testing has passed; and,
- PR has not received feedback for 7+ days.

gavindidrichsen replied:
Thanks @stevendanna. Will add these.

stevendanna commented on docs/dev/policy_documents/merging-and-promoting-conventions.md:
```diff
@@ -0,0 +1,40 @@
+## Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Delete this section?
```
I agree we can merge this and the previous section for this one.

gavindidrichsen replied:
Deleted.

stevendanna commented on docs/dev/policy_documents/merging-and-promoting-conventions.md:
```diff
@@ -0,0 +1,40 @@
+## Drawbacks
+[drawbacks]: #drawbacks
+
+What are the drawbacks if any?
```
Drawback include (1) less awareness amongst maintainers about what changes have occurred recently, (2) opportunity for bugs to enter that would have been caught in code review, (3) decreases pressure to improve the code-review process that has led to long delays.

gavindidrichsen replied:
Updating according to these comments

stevendanna commented on docs/dev/policy_documents/merging-and-promoting-conventions.md:
```diff
@@ -0,0 +1,40 @@
+
+What plan is eligable for the fast-track criteria:
+
+* Only identified 'veteran' core-plan contributor's have permission to use this fast-track policy.  This list will be agreed and specified in the MAINTAINERS.md
```
```suggestion
* Only identified 'veteran' core-plan contributors have permission to use this fast-track policy.  This list will be agreed and specified in the MAINTAINERS.md
```

stevendanna commented on docs/dev/policy_documents/merging-and-promoting-conventions.md:
```diff
@@ -0,0 +1,40 @@
+What plan is eligable for the fast-track criteria:
+
+* Only identified 'veteran' core-plan contributor's have permission to use this fast-track policy.  This list will be agreed and specified in the MAINTAINERS.md
+* This fasttrack will only apply to non-baseplan PRs and PRs that do not have a label 'DO-NOT-MERGE' or are otherwise marked for caution.
```
```suggestion
* This fast-track will only apply to non-base plan PRs and PRs that do not have a label 'DO-NOT-MERGE' or are otherwise marked for caution.
```

gavindidrichsen commented on MAINTAINERS.md:
```diff
@@ -45,3 +45,9 @@ To become a maintainer, open a pull request to this list.
+
+This section has been added for the benefit of those who contribute heavily to the core-plan's and are eligible for 'fastrack' promotion.  See [fastrack policy document](https://github.com/habitat-sh/core-plans/blob/master/docs/dev/policy_documents/merging-and-promoting-conventions.md)
+
+* [Graham Weldon](https://github.com/predominant)
```
@predominant: You're the first 'veteran'...

stevendanna replied:
@gavindidrichsen May I suggest that you add yourself to the list as well:

```
 > git shortlog -sne --since "3 months ago" | head -10
    66	Graham Weldon <graham@grahamweldon.com>
    32	Gavin Didrichsen <gavin.didrichsen@gmail.com>
    10	W. Duncan Fraser <duncan@wduncanfraser.com>
     7	Tim Smith <tsmith84@gmail.com>
     6	Scott Macfarlane <smacfarlane@chef.io>
     6	Christopher Maher <chris@mahercode.io>
     5	Ben Dang <me@bdang.it>
     4	mwrock <matt@mattwrock.com>
     4	Salim Afiune <afiune@chef.io>
     3	Steven Danna <steve@chef.io>
```

gavindidrichsen replied:
Done.

eeyun commented:
I'm also :+1: in concept. Would it make sense to just follow along with the chef oss practices pattern here and call folks that can merge approvers and then just add a note that they can approve their own PRs in the contexts outlined here?

gavindidrichsen commented:
> I'm also 👍 in concept. Would it make sense to just follow along with the chef oss practices pattern here and call folks that can merge approvers and then just add a note that they can approve their own PRs in the contexts outlined here?

@eeyun: Your suggestion makes sense.  

What about we do the following:
<img width="574" alt="image" src="https://user-images.githubusercontent.com/16652757/73083443-15973f80-3ec3-11ea-8024-248a497aa7bc.png">
* turn "Active Mainteners" into the "Approvers" list
* add a note that "Approvers can approve their own PRs according to policy in fasttrack document..."

Or should we simply turn 'Veteran' into 'Approvers' and add members?

gavindidrichsen commented on MAINTAINERS.md:
```diff
@@ -45,3 +45,10 @@ To become a maintainer, open a pull request to this list.
 * [Romain Sertelon](https://github.com/rsertelon)
 * [Scott Hain](https://github.com/scotthain)
+
+## Approvers
```
@eeyun, @stevendanna:  I've added this "approvers" section and adding myself along with Graham to the list.  I've also removed the term 'veteran' in favour of the 'approver' member, which is defined in the Chef Project Member handbook.  I've left the 'Active Members' section as-is (I didn't change this to be 'Approvers').  If 'Active Members' wish, they can raise a PR to get their names moved into the Approvers list.

gavindidrichsen replied:
Unless you (and @predominant) want me to make any more changes in the next couple days, I'll plan on merging this by the end of the week.  Thanks all.

gavindidrichsen commented:
Hi @stevendanna, @eeyun, @predominant .  I think I've addressed all comments/suggestions.  If all is ok would you mind sticking an approval on this so that I can get it merged into master?  Cheers.

predominant commented:
Approved 
![tenor-115095134](https://user-images.githubusercontent.com/24568/73314019-17208900-4270-11ea-9d02-75c4285d753b.gif)

gavindidrichsen commented:
> Overall, I'm +1 on the idea.
> 
> One minor point of concern is about the new 'veteran' maintainer status. I think the motivation here is that the reality of our maintainer list is that some maintainers (like me!) haven't been able to contribute to the project recently and might not be up to speed on the current state of core-plans. This is fair, but I wonder if we might just call this something like "Lead Maintainers" rather than 'veteran' since it is more about who is currently active in the project rather than tenure.. Another option might be to use the roles outlined here: https://github.com/chef/chef-oss-practices/blob/master/project-membership.md, but that seems like a big change that I'd like to avoid if possible.

Hi @stevendanna.  If I've addressed your comments fully, can you please resolve your request and approve?  Cheers!

stevendanna commented on docs/dev/policy_documents/merging-and-promoting-conventions.md:
```diff
@@ -0,0 +1,45 @@
+
+Our code review process is currently slow, frustrating core-plan maintainers who are donating their personal time or their employer's time to the project. One contributor (Graham Wilson, predominant) has hundreds of updates in his backlog, but has been holding back because of delays in the review process.
+
+If someone who contributes heavily to core plans has a pull request (PR) for a minor change open for 7+ days, the automated testing already in place and the small scope of the change makes the limited value of additional code-review small compared to the cost it is imposing on some maintainers.  In this situation, allowing 'aprover' maintainers to use their own judgement and immediately merge the ...
```
```suggestion
If someone who contributes heavily to core plans has a pull request (PR) for a minor change open for 7+ days, the automated testing already in place and the small scope of the change makes the value of additional code-review small compared to the cost it is imposing on some maintainers.  In this situation, allowing 'approver' maintainers to use their own judgement and immediately merge the PR will help ensure core-plan maintainers can make the most of the time they spend on this project.
```

gavindidrichsen commented:
> Sorry for the delay, I was on vacation. This looks good to me. I've left a minor inline correction that shouldn't block the merging of this PR.

Hi @stevendanna.  Thanks for the suggestions.  All commited now and merging.
